content/recipes/authoring/kata-snippets/submission-tests.md

```diff
@@ -0,0 +1,88 @@
+- **Provide both [fixed](#fixed-tests) and [random](#random-tests) tests** for the submit tests. Some test frameworks offer features like test case generators (`QuickCheck`, `ScalaCheck`), and when used, authors can choose different structures for their test suite, but they should still create test specifications covering all possible scenarios. See specific sections below for more details.
+- **Tests should not be susceptible to input mutation.** It should not be possible for the user's code to modify its input in a way which would influence the execution, result, or output of a test. If a reference solution is used, it should be called first, and the user solution should be called afterwards with the same input, or only pass a (deep) copy of the inputs to the user's solution. Potentially mutated input should not be used to compose any information presented to the user.
+- **Assertions should be placed near invocations of the user's solution**, preferably immediately following them. When several results are computed calling the user's function but the related assertions are only done further, what is printed to the console by the user isn't close to the related assertion in the output panel anymore. Debugging becomes difficult for the user and that makes the experience unpleasant.
+- **The test suite should not take a long time to run.** For kata which are not performance-oriented, the test suite should complete well below the time limit, ideally within 3 seconds. There's a time limit of 12 seconds (16 or 20 seconds for a few languages) for every attempt, but the generation and execution of test cases, including any calls to the user and reference solutions, should run well below this limit. Otherwise, there is a risk that tests will inconsistently time out, or some valid, but not optimally performant solutions will be rejected.
+- **Test cases should be independent**, they should not rely on any state passed from one test case to another.
+
+
+## Reference Solution
+
+Some test suites require a reference solution to generate the expected value(s) and compare them to those returned by the user solution. While such practice is common and sometimes necessary, there are things which have to be handled carefully:
+
+- **Avoid using a reference solution at all if possible.** Sometimes there's no other way, but quite often a reference solution is simply not necessary: test cases can be generated with the answer known upfront, which eliminates an entire class of potential problems associated with the use of reference solutions:
```

Comment:
```suggestion
- **Avoid using a reference solution at all if possible...** Sometimes there's no other way, but quite often a reference solution is simply not necessary and this can avoid a lot of hurdle. Test cases can effectively be generated with the answer known upfront, which eliminates an entire class of potential problems associated with the use of reference solutions, like the following:
```

adding emphasis with "..." caus the first thing we say is "here is what you need to be carefull about: just don't use it" x)
and avoiding double semi-colon imbrication

Reply:
Not sure what you mean "this can avoid a lot of hurdle"?

Reply:
changed to `troubles`... (to avoid some x) )

Reply:
changed to `trouble` (see comment further up)

Reply:
woops x)